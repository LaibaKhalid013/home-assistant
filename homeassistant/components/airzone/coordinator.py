"""The Airzone integration."""

from __future__ import annotations

from asyncio import timeout
from datetime import timedelta
import logging
from typing import Any

from aioairzone.const import AZD_NEW_SYSTEMS, AZD_NEW_ZONES
from aioairzone.exceptions import AirzoneError
from aioairzone.localapi import AirzoneLocalApi

from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import AIOAIRZONE_DEVICE_TIMEOUT_SEC, DOMAIN

SCAN_INTERVAL = timedelta(seconds=60)

_LOGGER = logging.getLogger(__name__)


class AirzoneUpdateCoordinator(DataUpdateCoordinator[dict[str, Any]]):
    """Class to manage fetching data from the Airzone device."""

    def __init__(self, hass: HomeAssistant, airzone: AirzoneLocalApi) -> None:
        """Initialize."""
        self.airzone = airzone

        super().__init__(
            hass,
            _LOGGER,
            name=DOMAIN,
            update_interval=SCAN_INTERVAL,
        )

    async def _async_check_new_devices(self, data: dict[str, Any]) -> None:
        """Reload entry if there are new devices."""
        if (
            len(data.get(AZD_NEW_SYSTEMS, [])) > 0
            or len(data.get(AZD_NEW_ZONES, [])) > 0
        ):
            assert self.config_entry
            self.hass.config_entries.async_schedule_reload(self.config_entry.entry_id)

    async def _async_update_data(self) -> dict[str, Any]:
        """Update data via library."""
        async with timeout(AIOAIRZONE_DEVICE_TIMEOUT_SEC):
            try:
                await self.airzone.update()
            except AirzoneError as error:
                raise UpdateFailed(error) from error
            data = self.airzone.data()
        await self._async_check_new_devices(data)
        return data
