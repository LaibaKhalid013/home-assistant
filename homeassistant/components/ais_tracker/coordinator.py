"""AIS tracker coordinator."""

from __future__ import annotations

import asyncio
import logging
from socket import AF_INET, SO_REUSEPORT, SOCK_DGRAM, SOL_SOCKET, socket

from pyais.stream import SocketStream

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PORT
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
from homeassistant.util.json import json_loads_object

from .const import CONF_MMSIS, DOMAIN

LOGGER = logging.getLogger(__name__)

type AisTrackerConfigEntry = ConfigEntry[AisTrackerCoordinator]


class UDPReceiver(SocketStream):
    """Re-implementation of pyais.stream.UDPReceiver.

    can be removed, as soon as https://github.com/M0r13n/pyais/pull/148 is merged.
    """

    _fobj: socket

    def __init__(self, host: str, port: int) -> None:
        """Initialize the UDP receiver."""
        sock: socket = socket(AF_INET, SOCK_DGRAM)
        sock.setsockopt(SOL_SOCKET, SO_REUSEPORT, 1)
        sock.bind((host, port))
        super().__init__(sock, preprocessor=None)

    def recv(self) -> bytes:
        """Receive data from socket."""
        return self._fobj.recvfrom(self.BUF_SIZE)[0]

    def close(self) -> None:
        """Close the UDP receiver."""
        self._fobj.close()


class AisTrackerCoordinator(
    DataUpdateCoordinator[dict[str, dict[str, float | int | str | None]]]
):
    """AIS tracker data update coordinator."""

    config_entry: AisTrackerConfigEntry

    def __init__(self, hass: HomeAssistant) -> None:
        """Initialize the AIS tracker coordinator."""
        super().__init__(
            hass,
            LOGGER,
            name=DOMAIN,
            update_interval=None,
        )
        self._receiver: UDPReceiver | None = None
        self._background_task: asyncio.Task | None = None
        self.data = {}

    async def async_setup(self) -> None:
        """Set up the AIS tracker coordinator."""

        self._receiver = UDPReceiver("", self.config_entry.data[CONF_PORT])

        async def async_ais_listerner():
            def ais_listerner():
                for msg_raw in self._receiver:
                    msg = json_loads_object(msg_raw.decode().to_json())
                    LOGGER.debug("received msg: %s", msg)
                    if (
                        msg.get("msg_type") in [1, 2, 3, 5]
                        and str(msg.get("mmsi")) in self.config_entry.data[CONF_MMSIS]
                    ):
                        self.hass.create_task(self.async_process_ais_message(msg))

            await self.hass.async_add_executor_job(ais_listerner)

        self._background_task = self.config_entry.async_create_background_task(
            hass=self.hass,
            target=async_ais_listerner(),
            name=f"{DOMAIN}_ais_listerner",
            eager_start=False,
        )

        self.config_entry.async_on_unload(self.async_teardown)

    async def async_teardown(self) -> None:
        """Tear down the AIS tracker coordinator."""
        if self._receiver is not None:
            self._receiver.close()
        if self._background_task is not None:
            self._background_task.cancel()

    async def async_process_ais_message(self, message: dict) -> None:
        """Process incoming ais message."""
        LOGGER.debug("process ais message: %s", message)
        if message.get("msg_type") in [1, 2, 3]:  # position reports
            data = {**self.data}
            data[str(message["mmsi"])] = message
            self.async_set_updated_data(data)

        elif message.get("msg_type") == 5:  # Static and voyage related data
            dreg = dr.async_get(self.hass)
            dreg.async_get_or_create(
                config_entry_id=self.config_entry.entry_id,
                identifiers={(DOMAIN, str(message["mmsi"]))},
                name=f"{message.get('shipname')} ({message.get('callsign')})",
                serial_number=str(message["mmsi"]),
                manufacturer="AIS tracker",
            )
