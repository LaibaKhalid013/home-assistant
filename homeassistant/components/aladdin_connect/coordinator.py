"""Define an object to coordinate fetching Aladdin Connect data."""

from datetime import timedelta
import logging

from genie_partner_sdk.client import AladdinConnectClient
from genie_partner_sdk.model import GarageDoor

from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

from .const import DOMAIN

_LOGGER = logging.getLogger(__name__)


class AladdinConnectCoordinator(DataUpdateCoordinator[None]):
    """Aladdin Connect Data Update Coordinator."""

    doors: list[GarageDoor] | None = None

    def __init__(self, hass: HomeAssistant, acc: AladdinConnectClient) -> None:
        """Initialize."""
        super().__init__(
            hass,
            logger=_LOGGER,
            name=DOMAIN,
            update_interval=timedelta(seconds=15),
        )
        self.acc = acc

    async def async_setup(self) -> None:
        """Fetch initial data."""
        self.doors = await self.acc.get_doors()

    async def _async_update_data(self) -> None:
        """Fetch data from API endpoint."""
        assert self.doors is not None
        for door in self.doors:
            await self.acc.update_door(door.device_id, door.door_number)
