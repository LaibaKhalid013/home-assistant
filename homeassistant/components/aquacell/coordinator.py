"""Coordinator to update data from Aquacell API."""

import asyncio
from datetime import timedelta
import logging

from aioaquacell import (
    AquacellApi,
    AquacellApiException,
    AuthenticationFailed,
    Softener,
)

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryError
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import CONF_REFRESH_TOKEN, UPDATE_INTERVAL

_LOGGER = logging.getLogger(__name__)


class AquacellCoordinator(DataUpdateCoordinator[dict[str, Softener]]):
    """My aquacell coordinator."""

    config_entry: ConfigEntry

    def __init__(self, hass: HomeAssistant, aquacell_api: AquacellApi) -> None:
        """Initialize coordinator."""
        super().__init__(
            hass,
            _LOGGER,
            name="Aquacell Coordinator",
            update_interval=timedelta(seconds=UPDATE_INTERVAL),
        )

        self.refresh_token = self.config_entry.data[CONF_REFRESH_TOKEN]
        self.email = self.config_entry.data[CONF_EMAIL]
        self.password = self.config_entry.data[CONF_PASSWORD]
        self.aquacell_api = aquacell_api

    async def _async_update_data(self) -> dict[str, Softener]:
        """Fetch data from API endpoint.

        This is the place to pre-process the data to lookup tables
        so entities can quickly look up their data.
        """

        async with asyncio.timeout(10):
            _LOGGER.debug("Logging in using: %s", self.refresh_token)

            try:
                await self.aquacell_api.authenticate_refresh(self.refresh_token)
            except AuthenticationFailed as err:
                _LOGGER.debug(
                    "Authentication using refresh token failed due to: %s", err
                )
                _LOGGER.debug("Attempting to renew refresh token")
                refresh_token = await self.aquacell_api.authenticate(
                    self.email, self.password
                )
                self.refresh_token = refresh_token
                self._update_config_entry_refresh_token()

            _LOGGER.debug("Logged in, new token: %s", self.aquacell_api.id_token)
        try:
            softeners = await self.aquacell_api.get_all_softeners()
        except AuthenticationFailed as err:
            raise ConfigEntryError from err
        except AquacellApiException as err:
            raise UpdateFailed(f"Error communicating with API: {err}") from err

        return {softener.dsn: softener for softener in softeners}

    def _update_config_entry_refresh_token(self) -> None:
        """Update or delete the refresh_token in the Config Entry."""
        data = {
            **self.config_entry.data,
            CONF_REFRESH_TOKEN: self.refresh_token,
        }

        self.hass.config_entries.async_update_entry(self.config_entry, data=data)
