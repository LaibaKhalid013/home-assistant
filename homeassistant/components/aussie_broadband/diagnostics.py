"""Provides diagnostics for Aussie Broadband."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from homeassistant.components.diagnostics import async_redact_data

from .const import DOMAIN

if TYPE_CHECKING:
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant

TO_REDACT = ["address", "ipAddresses", "description", "discounts", "coordinator"]


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    return {
        "services": [
            {
                "service": async_redact_data(service, TO_REDACT),
                "usage": async_redact_data(service["coordinator"].data, ["historical"]),
            }
            for service in hass.data[DOMAIN][config_entry.entry_id]["services"]
        ]
    }
