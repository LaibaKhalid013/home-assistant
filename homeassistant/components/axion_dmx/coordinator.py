"""Coordinator for Axion Lighting integration."""

from datetime import timedelta
from typing import Any

from requests.exceptions import RequestException

from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import _LOGGER


class AxionDataUpdateCoordinator(DataUpdateCoordinator):
    """Custom coordinator for Axion Lighting integration."""

    def __init__(self, hass: HomeAssistant, api: Any, channel: int) -> None:
        """Initialize the Axion data coordinator."""
        self.api = api
        self.channel = channel
        super().__init__(
            hass,
            _LOGGER,
            name="Axion Light",
            update_method=self._async_update_data,
            update_interval=timedelta(seconds=5),
        )

    async def _async_update_data(self) -> Any:
        """Fetch data from the API endpoint."""
        try:
            return await self.api.get_level(self.channel)
        except RequestException as err:
            raise UpdateFailed("Error communicating with API") from err
