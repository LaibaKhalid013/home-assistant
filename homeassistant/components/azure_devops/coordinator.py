"""Define the Azure DevOps DataUpdateCoordinator."""

from datetime import timedelta
import logging
from typing import Final

from aioazuredevops.builds import DevOpsBuild
from aioazuredevops.client import DevOpsClient
from aioazuredevops.core import DevOpsProject
import aiohttp

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryAuthFailed
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import CONF_ORG, CONF_PROJECT, DOMAIN
from .data import AzureDevOpsData

BUILDS_QUERY: Final = "?queryOrder=queueTimeDescending&maxBuildsPerDefinition=1"


class AzureDevOpsDataUpdateCoordinator(DataUpdateCoordinator[AzureDevOpsData]):
    """Class to manage and fetch Azure DevOps data."""

    def __init__(
        self,
        hass: HomeAssistant,
        logger: logging.Logger,
        *,
        entry: ConfigEntry,
    ) -> None:
        """Initialize global Azure DevOps data updater."""
        self.title = entry.title

        super().__init__(
            hass=hass,
            logger=logger,
            name=DOMAIN,
            update_interval=timedelta(seconds=300),
        )

        self.client = DevOpsClient(session=async_get_clientsession(hass))
        self.organization = entry.data[CONF_ORG]
        self.project_name = entry.data[CONF_PROJECT]

    async def authorize(
        self,
        personal_access_token: str,
    ) -> None:
        """Authorize with Azure DevOps."""
        await self.client.authorize(
            personal_access_token,
            self.organization,
        )
        if not self.client.authorized:
            raise ConfigEntryAuthFailed(
                "Could not authorize with Azure DevOps. You will need to update your"
                " token"
            )

    async def _get_project(
        self,
        project: str,
    ) -> DevOpsProject | None:
        """Get the project."""
        return await self.client.get_project(
            self.organization,
            project,
        )

    async def _get_builds(self, project_name: str) -> list[DevOpsBuild] | None:
        """Get the builds."""
        return await self.client.get_builds(
            self.organization,
            project_name,
            BUILDS_QUERY,
        )

    async def _async_update_data(self) -> AzureDevOpsData:
        """Fetch data from Azure DevOps."""
        try:
            # Get the project if we haven't already
            if self.data is None or (project := self.data.project) is None:
                project = await self._get_project(self.project_name)
                if project is None:
                    raise UpdateFailed("No project found")

            # Get the builds from the project
            builds = await self._get_builds(project.name)
        except aiohttp.ClientError as exception:
            raise UpdateFailed from exception

        if builds is None:
            raise UpdateFailed("No builds found")

        return AzureDevOpsData(
            organization=self.organization,
            project=project,
            builds=builds,
        )
