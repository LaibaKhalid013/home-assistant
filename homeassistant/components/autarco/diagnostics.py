"""Support for the Autarco diagnostics."""

from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
from homeassistant.core import HomeAssistant

from . import AutarcoConfigEntry, AutarcoDataUpdateCoordinator

TO_REDACT = {CONF_EMAIL, CONF_PASSWORD}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: AutarcoConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    autarco_data: list[AutarcoDataUpdateCoordinator] = config_entry.runtime_data

    return {
        "config_entry_data": async_redact_data(dict(config_entry.data), TO_REDACT),
        "sites_data": [
            {
                "id": coordinator.site.site_id,
                "name": coordinator.site.system_name,
                "health": coordinator.site.health,
                "solar": {
                    "power_production": coordinator.data.solar.power_production,
                    "energy_production_today": coordinator.data.solar.energy_production_today,
                    "energy_production_month": coordinator.data.solar.energy_production_month,
                    "energy_production_total": coordinator.data.solar.energy_production_total,
                },
            }
            for coordinator in autarco_data
        ],
    }
