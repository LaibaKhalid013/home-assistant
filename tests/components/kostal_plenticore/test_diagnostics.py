"""Test Kostal Plenticore diagnostics."""

from pykoplenti import SettingsData
from syrupy import SnapshotAssertion

from homeassistant.components.kostal_plenticore.coordinator import Plenticore
from homeassistant.core import HomeAssistant

from tests.common import MockConfigEntry
from tests.components.diagnostics import snapshot_get_diagnostics_for_config_entry
from tests.typing import ClientSessionGenerator


async def test_entry_diagnostics(
    hass: HomeAssistant,
    hass_client: ClientSessionGenerator,
    mock_plenticore: Plenticore,
    init_integration: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test config entry diagnostics."""

    # set some test process and settings data for the diagnostics output
    mock_plenticore.client.get_process_data.return_value = {
        "devices:local": ["HomeGrid_P", "HomePv_P"]
    }

    mock_plenticore.client.get_settings.return_value = {
        "devices:local": [
            SettingsData(
                min="5",
                max="100",
                default=None,
                access="readwrite",
                unit="%",
                id="Battery:MinSoc",
                type="byte",
            )
        ]
    }

    await snapshot_get_diagnostics_for_config_entry(
        hass, hass_client, init_integration, snapshot
    )
