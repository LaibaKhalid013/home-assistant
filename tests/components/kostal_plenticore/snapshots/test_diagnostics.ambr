# serializer version: 1
# name: test_entry_diagnostics
  dict({
    'client': dict({
      'available_process_data': dict({
        'devices:local': list([
          'HomeGrid_P',
          'HomePv_P',
        ]),
      }),
      'available_settings_data': dict({
        'devices:local': list([
          "min='5' max='100' default=None access='readwrite' unit='%' id='Battery:MinSoc' type='byte'",
        ]),
      }),
      'me': "is_locked=False is_active=True is_authenticated=True permissions=[] is_anonymous=False role='USER'",
      'version': "api_version='0.2.0' hostname='scb' name='PUCK RESTful API' sw_version='01.16.05025'",
    }),
    'config_entry': dict({
      'data': dict({
        'host': '192.168.1.2',
        'password': '**REDACTED**',
      }),
      'disabled_by': None,
      'domain': 'kostal_plenticore',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'title': 'scb',
      'unique_id': None,
      'version': 1,
    }),
    'device': dict({
      'configuration_url': 'http://192.168.1.2',
      'identifiers': '**REDACTED**',
      'manufacturer': 'Kostal',
      'model': 'PLENTICORE plus 10',
      'name': 'scb',
      'sw_version': 'IOC: 01.45 MC: 01.46',
    }),
  })
# ---
